File: backend/app/models/messages.py
from __future__ import annotations

from datetime import datetime, timezone
from enum import Enum
from typing import Any
from uuid import uuid4

from pydantic import BaseModel, Field


class MessageRole(str, Enum):
    USER = "user"
    ASSISTANT = "assistant"
    SYSTEM = "system"


# -- Content types -------------------------------------------------------

class TextContent(BaseModel):
    type: str = "text"
    text: str


class OptionItem(BaseModel):
    id: str
    label: str
    description: str | None = None


class OptionsContent(BaseModel):
    type: str = "options"
    question: str
    question_id: str
    options: list[OptionItem]
    allow_multiple: bool = False
    allow_freeform: bool = True


class ProgressContent(BaseModel):
    type: str = "progress"
    phase: str
    message: str


class ArchitectureContent(BaseModel):
    type: str = "architecture"
    summary: dict[str, Any]


ContentType = TextContent | OptionsContent | ProgressContent | ArchitectureContent


# -- Chat message ---------------------------------------------------------

class ChatMessage(BaseModel):
    id: str = Field(default_factory=lambda: str(uuid4()))
    role: MessageRole
    content: list[ContentType]
    timestamp: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))


# -- WebSocket protocol ---------------------------------------------------

class WSMessageType(str, Enum):
    # Client -> Server
    USER_MESSAGE = "user_message"
    OPTION_SELECTED = "option_selected"
    GENERATE_CODE = "generate_code"

    # Server -> Client
    ASSISTANT_CHUNK = "assistant_chunk"
    ASSISTANT_COMPLETE = "assistant_complete"
    SYSTEM_MESSAGE = "system_message"
    OPTIONS_PROMPT = "options_prompt"
    ARCHITECTURE_READY = "architecture_ready"
    GENERATION_PROGRESS = "generation_progress"
    GENERATION_COMPLETE = "generation_complete"
    DEPLOYMENT_PROGRESS = "deployment_progress"
    DEPLOYMENT_COMPLETE = "deployment_complete"
    ERROR = "error"


class WSFrame(BaseModel):
    type: WSMessageType
    session_id: str
    payload: dict[str, Any]
    timestamp: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))

File: frontend/src/types/messages.ts
// -- Message types matching backend Pydantic models --

export type MessageRole = "user" | "assistant" | "system";

export type ConversationPhase =
  | "initial"
  | "understanding"
  | "clarifying"
  | "designing"
  | "complete"
  | "generating"
  | "deploying";

// Content types
export interface TextContent {
  type: "text";
  text: string;
}

export interface OptionItem {
  id: string;
  label: string;
  description?: string;
}

export interface OptionsContent {
  type: "options";
  question: string;
  question_id: string;
  options: OptionItem[];
  allow_multiple: boolean;
  allow_freeform: boolean;
}

export interface ProgressContent {
  type: "progress";
  phase: ConversationPhase;
  message: string;
}

export interface ArchitectureContent {
  type: "architecture";
  summary: Record<string, unknown>;
}

export type ContentBlock =
  | TextContent
  | OptionsContent
  | ProgressContent
  | ArchitectureContent;

// Chat message
export interface ChatMessage {
  id: string;
  role: MessageRole;
  content: ContentBlock[];
  timestamp: string;
}

// WebSocket protocol
export type WSMessageType =
  | "user_message"
  | "option_selected"
  | "generate_code"
  | "assistant_chunk"
  | "assistant_complete"
  | "system_message"
  | "options_prompt"
  | "architecture_ready"
  | "generation_progress"
  | "generation_complete"
  | "deployment_progress"
  | "deployment_complete"
  | "error";

export interface WSFrame {
  type: WSMessageType;
  session_id: string;
  payload: Record<string, unknown>;
  timestamp: string;
}

// Architecture types (for display)
export interface ToolDefinition {
  name: string;
  file_name: string;
  description: string;
  parameters: ToolParameter[];
  return_description: string;
  annotations: {
    read_only_hint: boolean;
    destructive_hint: boolean;
    idempotent_hint: boolean;
    open_world_hint: boolean;
  };
  external_api?: string;
  error_scenarios: string[];
}

export interface ToolParameter {
  name: string;
  type: string;
  description: string;
  required: boolean;
  default?: string;
  enum_values?: string[];
}

export interface MCPArchitecture {
  server_name: string;
  server_description: string;
  server_version: string;
  language: string;
  language_reasoning: string;
  enable_tools: boolean;
  enable_resources: boolean;
  enable_prompts: boolean;
  tools: ToolDefinition[];
  extra_dependencies: { package: string; reason: string }[];
  env_vars: { name: string; description: string; required: boolean; example: string }[];
  auth_setup?: {
    type: string;
    description: string;
    env_var_names: string[];
  };
}

// Deployment info
export interface DeploymentInfo {
  status: "running" | "failed";
  server_url?: string;
  sse_url?: string;
  port?: number;
  client_config?: Record<string, unknown>;
  client_config_json?: string;
  claude_desktop_config?: string;
  cursor_config?: string;
  generic_config?: string;
  error?: string;
}
